File: labs/01-hello-model/typescript/solution/05_same_code_other_model.mjs
// Lab 1 · step 5 — same code, different model. The point of the day in one script.
//
// Runs one prompt and one tool round-trip against every endpoint it can find in labs/.env:
//   the workshop endpoint (GENAI_MODEL), a second vendor behind the same key (GENAI_MODEL_2),
//   local Ollama (OLLAMA_BASE_URL / OLLAMA_MODEL) and LM Studio (LMSTUDIO_BASE_URL) if they are running.
import { existsSync } from "node:fs";
import { dirname, join } from "node:path";
import OpenAI from "openai";

let LABS = import.meta.dirname;                     // labs/.env (real environment variables win)
while (!existsSync(join(LABS, "_tools", "make_starters.py")) && LABS !== dirname(LABS)) LABS = dirname(LABS);
// labs/.env, or one .env at the repo root: load every .env from labs/ upward (never overwrites) - nearest wins
for (let dir = LABS; ; dir = dirname(dir)) {
  if (existsSync(join(dir, ".env"))) process.loadEnvFile(join(dir, ".env"));
  if (dirname(dir) === dir) break;
}

const { GENAI_BASE_URL: BASE, GENAI_API_KEY: KEY, GENAI_MODEL: MODEL } = process.env;
if (!BASE || !KEY || !MODEL) throw new Error("Set GENAI_BASE_URL, GENAI_API_KEY and GENAI_MODEL in labs/.env (see .env.example)");
const HEADERS = (process.env.GENAI_KEY_HEADER ?? "api-key") === "api-key" ? { "api-key": KEY } : undefined;

const env = process.env;
const [EFFORT, LOCAL_EFFORT] = [env.GENAI_REASONING_EFFORT, env.OLLAMA_REASONING_EFFORT];
const targets = [["workshop endpoint", BASE, KEY, HEADERS, MODEL]];
if (env.GENAI_MODEL_2) targets.push(["2nd vendor, same key", BASE, KEY, HEADERS, env.GENAI_MODEL_2]);
if (env.OLLAMA_BASE_URL && env.OLLAMA_MODEL) targets.push(["local Ollama", env.OLLAMA_BASE_URL, "ollama", undefined, env.OLLAMA_MODEL]);
if (env.LMSTUDIO_BASE_URL) targets.push(["local LM Studio", env.LMSTUDIO_BASE_URL, "lm-studio", undefined, env.LMSTUDIO_MODEL ?? ""]);

const TOOL = { type: "function", function: { name: "get_weather", description: "Current weather for a city.",
  parameters: { type: "object", properties: { city: { type: "string" } }, required: ["city"] } } };
const seconds = (t0) => ((performance.now() - t0) / 1000).toFixed(1).padStart(5);

for (let [label, base, key, headers, model] of targets) {
  const client = new OpenAI({ baseURL: base, apiKey: key, defaultHeaders: headers, timeout: 180_000, maxRetries: 0 });
  const effort = label.startsWith("local") ? LOCAL_EFFORT : (label === "workshop endpoint" ? EFFORT : undefined);
  const extra = effort ? { reasoning_effort: effort } : {};
  try {
    if (!model) model = (await client.models.list()).data[0].id;   // LM Studio: take whatever model is loaded
    let t0 = performance.now();
    const r = await client.chat.completions.create({ model, ...extra, messages: [
      { role: "user", content: "Explain retrieval-augmented generation to a developer in one sentence." }] });
    const chatS = seconds(t0);
    t0 = performance.now();
    const r2 = await client.chat.completions.create({ model, tools: [TOOL], ...extra, messages: [
      { role: "user", content: "What's the weather in Mannheim?" }] });
    const calls = r2.choices[0].message.tool_calls ?? [];
    const toolS = seconds(t0);
    console.log(`\n== ${label}: ${model}  (${base})`);
    console.log(`   chat  ${chatS}s  ${r.usage?.completion_tokens ?? "?"} tokens: ` +
      `${(r.choices[0].message.content ?? "").trim().slice(0, 160)}`);
    console.log(`   tools ${toolS}s  ` + (calls.map((c) => `${c.function.name}(${c.function.arguments})`).join(", ")
      || `no tool call -> ${(r2.choices[0].message.content ?? "").slice(0, 80)}`));
  } catch (e) {                                     // a missing local runtime must not stop the comparison
    console.log(`\n== ${label}: ${model || "?"}  skipped (${e.constructor.name}: ${String(e.message).slice(0, 120)})`);
  }
}

// What to notice: identical code, different answers, latency and tool-calling habits.
// Independence = being able to make this switch in one line — and knowing what you lose or gain when you do.

File: labs/01-hello-model/python/solution/05_same_code_other_model.py
"""Lab 1 · step 5 — same code, different model. The point of the day in one script.

Runs one prompt and one tool round-trip against every endpoint it can find in labs/.env:
  the workshop endpoint (GENAI_MODEL), a second vendor behind the same key (GENAI_MODEL_2),
  local Ollama (OLLAMA_BASE_URL / OLLAMA_MODEL) and LM Studio (LMSTUDIO_BASE_URL) if they are running.
"""
import json
import os
import time

from dotenv import find_dotenv, load_dotenv
from pathlib import Path as _Path
from openai import OpenAI

_here = _Path(globals().get("__file__", ".")).resolve()   # no __file__ in a notebook: start at the cwd
for _folder in [*_here.parents, _Path.cwd(), *_Path.cwd().parents]:
    if (_folder / ".env").is_file():   # every .env from this file upward: the nearest one wins,
        load_dotenv(_folder / ".env")  # so labs/.env or one .env at the repo root both work
BASE, KEY = os.environ["GENAI_BASE_URL"], os.environ["GENAI_API_KEY"]
HEADERS = {"api-key": KEY} if os.getenv("GENAI_KEY_HEADER", "api-key") == "api-key" else None

EFFORT, LOCAL_EFFORT = os.getenv("GENAI_REASONING_EFFORT"), os.getenv("OLLAMA_REASONING_EFFORT")
targets = [("workshop endpoint", BASE, KEY, HEADERS, os.environ["GENAI_MODEL"])]
if os.getenv("GENAI_MODEL_2"):
    targets.append(("2nd vendor, same key", BASE, KEY, HEADERS, os.environ["GENAI_MODEL_2"]))
if os.getenv("OLLAMA_BASE_URL") and os.getenv("OLLAMA_MODEL"):
    targets.append(("local Ollama", os.environ["OLLAMA_BASE_URL"], "ollama", None, os.environ["OLLAMA_MODEL"]))
if os.getenv("LMSTUDIO_BASE_URL"):
    targets.append(("local LM Studio", os.environ["LMSTUDIO_BASE_URL"], "lm-studio", None, os.getenv("LMSTUDIO_MODEL", "")))

TOOL = {"type": "function", "function": {"name": "get_weather", "description": "Current weather for a city.",
        "parameters": {"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]}}}

for label, base, key, headers, model in targets:
    client = OpenAI(base_url=base, api_key=key, default_headers=headers, timeout=180, max_retries=0)
    effort = LOCAL_EFFORT if label.startswith("local") else (EFFORT if label == "workshop endpoint" else None)
    extra = {"reasoning_effort": effort} if effort else {}
    try:
        if not model:                                   # LM Studio: take whatever model is loaded
            model = client.models.list().data[0].id
        t0 = time.perf_counter()
        r = client.chat.completions.create(model=model, **extra, messages=[
            {"role": "user", "content": "Explain retrieval-augmented generation to a developer in one sentence."}])
        chat_s = time.perf_counter() - t0
        t0 = time.perf_counter()
        r2 = client.chat.completions.create(model=model, tools=[TOOL], **extra, messages=[
            {"role": "user", "content": "What's the weather in Mannheim?"}])
        calls = r2.choices[0].message.tool_calls or []
        tool_s = time.perf_counter() - t0
        print(f"\n== {label}: {model}  ({base})")
        print(f"   chat  {chat_s:5.1f}s  {r.usage.completion_tokens if r.usage else '?'} tokens: "
              f"{(r.choices[0].message.content or '').strip()[:160]}")
        print(f"   tools {tool_s:5.1f}s  " + (", ".join(f"{c.function.name}({json.loads(c.function.arguments)})" for c in calls)
                                            or f"no tool call -> {(r2.choices[0].message.content or '')[:80]}"))
    except Exception as e:                              # a missing local runtime must not stop the comparison
        print(f"\n== {label}: {model or '?'}  skipped ({type(e).__name__}: {str(e)[:120]})")

# What to notice: identical code, different answers, latency and tool-calling habits.
# Independence = being able to make this switch in one line — and knowing what you lose or gain when you do.
